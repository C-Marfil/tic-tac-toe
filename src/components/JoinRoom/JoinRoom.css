@import url('https://fonts.googleapis.com/css2?family=Odibee+Sans&display=swap');
.ChakraProvider {
  font-family: 'Odibee Sans', cursive;
  margin: 0;
}

.joinRoomContainer {
  margin-top: 135px;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  flex-flow: column wrap;
  padding-top: 65px;
  font-family: 'Odibee Sans', cursive;
}

.joinRoomContainer h3 {
  font-size: 2.5rem;
  margin-bottom: 1rem;
  font-family: 'Odibee Sans', cursive;
}

.joinRoomContainer input {
  width: 210px;
  height: 40px;
  margin: 7px;
  border: 2px solid #3e92cc;
  border-radius: 5px;
  padding: 5px;
  font-size: 16px;
}

.joinRoomContainer button {
  width: 225px;
  height: 50px;
  margin: 7px;
  border: none;
  border-radius: 5px;
  padding: 5px;
  font-size: 16px;
  border-color: 2px solid #ff6700;
  cursor: pointer;
}

.joinRoomContainer button:hover {
  background: #ff6700;
}

.joinRoomContainer input:focus {
  outline-color: #ff6700;
}

.roomError {
  font-size: 20px;
  color: #ffff;
}

/* .main-join-room{
display: flex;
flex-direction: column;
align-items: center;
justify-content: center;
}

.red {
  display: flex;
  flex-direction: row;
  justify-content: center;
  padding-top: 50px;
}

.refresh-button {
  font-size: 20px;
  color: black;
  background-color: black;
} */

/* .saveButton {
  display: flex;
  flex-direction: row;
  border: black solid 1px;
} */
/* .joinRoomModal-container {
    height: 100%;
    width: 100%;
    backdrop-filter: blur(3px);
  
    display: grid;
    place-content: center;
  
    position: fixed;
  }
  
  .joinRoomModal-container::after {
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    opacity: 0.4;
    z-index: -1;
    background: var(--primary-color);
  }
  
  .joinRoomModal-card {
    display: flex;
    flex-direction: column;
    gap: 1rem;
  
    padding: 1rem;
  }
  
  .joinRoomModal-card-title {
    color: var(--secondary-color);
    text-align: center;
  }
  
  .joinRoomModal-card-input {
    font-size: 2rem;
    border: 3px solid var(--secondary-color);
    background-color: var(--primary-color);
    outline: none;
    padding: 1rem;
    border-radius: 16px;
    color: var(--secondary-color);
  }
  
  .joinRoomModal-card-button {
    background-color: var(--secondary-color);
    border: 3px solid var(--primary-color);
    border-radius: 16px;
    font-size: 2rem;
    padding: 1rem;
    color: var(--primary-color);
  }
  
  .joinRoomModal-card-button:hover {
    cursor: pointer;
  } */
