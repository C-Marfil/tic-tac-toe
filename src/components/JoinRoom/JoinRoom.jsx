/* eslint-disable react-hooks/exhaustive-deps */
/* eslint-disable no-console */
/* eslint-disable no-alert */
import React, { useState, useEffect } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import { Button } from "@chakra-ui/react";
import PropTypes from "prop-types";
import "./JoinRoom.css";

const JoinRoom = ({ setRoomCode, socket, roomCode }) => {
  const [roomCodeInput, setRoomCodeInput] = useState(null);
  const [rooms, setRooms] = useState([]);
  const [error, setError] = useState(false);

  const navigate = useNavigate();
  const location = useLocation();

  const lobby = JSON.parse(localStorage.getItem("lobby"));

  const handleSave = (e) => {
    e.preventDefault();
    if (roomCodeInput === null) {
      setError(true);
    }
    if (roomCodeInput) {
      socket.emit("update-lobby", roomCodeInput);
      console.log("...updating Lobby");
      setRoomCode(roomCodeInput);
      setError(false);
      navigate(`/room${roomCodeInput}`);
    }
  };

  socket.on("rawRoomsString-incoming", (rawRoomsString) => {
    console.log("...setting lobby in localStorage");
    localStorage.setItem("lobby", rawRoomsString);
    setRooms(lobby);
  });

  const handleRoomClick = (e) => {
    const buttonID = e.target.id;
    setRoomCode(buttonID);
    socket.emit("update-lobby", roomCode);
    console.log("...updating Lobby");
    navigate(`/room${e.target.id}`);
  };

  const handleRefresh = () => {
    if (lobby) {
      setRooms(() => lobby);
      console.log("this is lobby at refresh", lobby);
    }
  };

  useEffect(() => {
    if (lobby) {
      setRooms(() => lobby);
      console.log("this is lobby at location useEffect", lobby);
    }
  }, [location, roomCodeInput]);

  useEffect(() => {
    if (roomCode !== null) {
      socket.emit("update-lobby", roomCode);
      console.log("...updating Lobby");
    }
  }, [roomCode]);

  return (
    <div className="joinRoomContainer">
      <h1 className="joinRoomCardTitle">Enter a room code</h1>
      <div>
        {/* <button type="button" onClick={handleRefresh}>
            Refresh Rooms
          </button> */}
        <form>
          <input
            type="text"
            aria-label="username-input"
            placeholder="eg: 14"
            value={roomCodeInput}
            onChange={(e) => setRoomCodeInput(e.target.value)}
          />
        </form>
        <Button
          colorScheme="brand"
          className="saveButton"
          variant="solid"
          type="submit"
          onClick={(e) => handleSave(e)}
        >
          Save & Join
        </Button>
        {/* <input
            className="roomcode-input"
            aria-label="roomcode-input"
            type="number"
            value={roomCodeInput}
            placeholder="eg: 1212"
            onChange={(e) => setRoomCodeInput(e.target.value)}
            />
            <button
            type="submit"
            onClick={(e) => handleSave(e)}
            className="JoinRoom-card-button"
            >
            Save
          </button> */}
      </div>
      {rooms.map((room) => {
        return (
          <Button
            colorScheme="brand"
            variant="solid"
            id={room}
            onClick={(e) => {
              handleRoomClick(e);
            }}
          >
            Join Room {room}
          </Button>
        );
      })}
      <Button
        className="refresh-button"
        colorScheme="brand"
        variant="solid"
        onClick={handleRefresh}
      >
        Refresh Rooms
      </Button>
      {error && (
        <p className="roomError">Please enter a room code to create a room</p>
      )}
    </div>
  );
};

JoinRoom.defaultProps = {
  roomCode: null,
};

JoinRoom.propTypes = {
  setRoomCode: PropTypes.func.isRequired,
  roomCode: PropTypes.string,
  socket: PropTypes.shape({
    emit: PropTypes.func.isRequired,
    on: PropTypes.func.isRequired,
  }).isRequired,
};

export default JoinRoom;
