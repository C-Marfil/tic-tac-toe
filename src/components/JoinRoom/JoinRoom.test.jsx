import React from "react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import { render, screen, fireEvent } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import JoinRoom from "./JoinRoom";

describe("JoinRoom, children and Routing render/work properly", () => {
  const rooms = [];
  const socket = {
    on: jest.fn(rooms.push(13)),
    off: jest.fn(),
    emit: jest.fn(),
  };

  const validProps = {
    socket,
    setRoomCode: jest.fn(),
    roomCode: "",
  };

  const setup = () => {
    const utils = render(
      <JoinRoom
        socket={validProps.socket}
        setRoomCode={validProps.setRoomCode}
        roomCode={validProps.roomCode}
      />,
      { wrapper: BrowserRouter }
    );
    const input = screen.getByLabelText("roomcode-input");
    return {
      input,
      ...utils,
    };
  };

  it("Snapshot test", () => {
    const { asFragment } = render(
      <JoinRoom
        socket={validProps.socket}
        setRoomCode={validProps.setRoomCode}
        roomCode={validProps.roomCode}
      />,
      { wrapper: BrowserRouter }
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("tests the component renders", () => {
    render(
      <JoinRoom
        socket={validProps.socket}
        setRoomCode={validProps.setRoomCode}
        roomCode={validProps.roomCode}
      />,
      { wrapper: BrowserRouter }
    );

    expect(screen.getByText(/Enter a room code/i).toBeInTheDocument);
    expect(screen.getByText(/Refresh Rooms/i).toBeInTheDocument);
    expect(screen.getByText(/Save/i).toBeInTheDocument);
  });

  it("input coderooms and throws error if input empty", () => {
    const { input } = setup();

    userEvent.click(screen.getByText(/Save/i));
    expect(
      screen.getByText(/Please enter a room code to create a room/i)
    ).toBeInTheDocument();

    fireEvent.change(input, { target: { value: "13" } });
    expect(input.value).toBe("13");

    userEvent.click(screen.getByText(/Save/i));
    expect(() =>
      screen.getByText(/Please enter a room code to create a room/i)
    ).toThrow("Unable to find an element");
  });
});
