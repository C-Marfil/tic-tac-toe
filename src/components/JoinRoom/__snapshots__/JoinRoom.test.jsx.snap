// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinRoom, children and Routing render/work properly Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="joinRoomContainer"
  >
    <h1
      class="joinRoomCardTitle"
    >
      Enter a room code
    </h1>
    <div>
      <form>
        <input
          aria-label="roomcode-input"
          class="roomcode--input"
          placeholder="eg: 14"
          type="text"
          value=""
        />
      </form>
      <button
        class="chakra-button saveButton css-4xx2wk"
        type="submit"
      >
        Save & Join
      </button>
      <button
        class="chakra-button refresh-button css-4xx2wk"
        type="button"
      >
        Refresh Rooms
      </button>
    </div>
  </div>
</DocumentFragment>
`;
